File: BE/src/main/java/com/react/config/SecurityConfig.java
package com.react.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.json.JsonMapper;
import com.react.auth.controller.JwtAuthenticationFilterController;
import com.react.auth.service.JwtService;
import com.react.exception.CustomAccessDeniedHandler;
import com.react.exception.CustomAuthenticationEntryPoint;
import com.react.user.repository.IUserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.List;

@Configuration
@RequiredArgsConstructor
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
public class SecurityConfig {
    private final IUserRepository  userRepository;
    private final CustomAccessDeniedHandler accessDeniedHandler;
    private final CustomAuthenticationEntryPoint authenticationEntryPoint;

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public AuthenticationManager authenticationManager(
            AuthenticationConfiguration configuration
    ) throws Exception {
        return configuration.getAuthenticationManager();
    }

    @Bean
    public UserDetailsService userDetailsService() {
        return email -> userRepository.findByEmail(email)
                .orElseThrow(() ->
                        new UsernameNotFoundException(
                                "User not found with email " + email
                        )
                );
    }

    @Bean
    public SecurityFilterChain securityFilterChain(
            HttpSecurity http,
            JwtAuthenticationFilterController jwtFilter
    ) throws Exception {


        http
                .cors(cors -> {})
                .csrf(csrf -> csrf.disable())
                .sessionManagement(session ->
                        session.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
                )
                .authorizeHttpRequests(auth -> auth
                        //các đường dẫn sau được đi qua bước phân quyền nhưng vẫn phải đi qua jwtFilter
                        .requestMatchers("/api/auth/**", "/error", "/graphql", "/ws/**", "/ws-sockjs/**", "/chat/**", "/api/test/**")
                        .permitAll()
                        // các đường dẫn khác phải đi qua được jwtFilter và setAuthentication được thì mới qua được day
                        .anyRequest().authenticated()
                )
                // UsernamePasswordAuthenticationFilter chỉ bắt các rq có /login các rq khác kh bắt và cho đi tiếp
                .addFilterBefore(jwtFilter, UsernamePasswordAuthenticationFilter.class)
                .exceptionHandling(ex -> ex
                        .accessDeniedHandler(accessDeniedHandler)
                        .authenticationEntryPoint(authenticationEntryPoint)
                );

        return http.build();
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration config = new CorsConfiguration();

        config.setAllowedOrigins(List.of("http://localhost:5173"));
        config.setAllowedMethods(List.of("GET", "POST", "PUT", "DELETE", "OPTIONS"));
        config.setAllowedHeaders(List.of("*"));
        config.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source =
                new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", config);

        return source;
    }
}

File: BaseSpringSecurity/src/main/kotlin/security/SecurityConfig.kt
package com.security

import com.auth.controller.JwtAuthenticationFilterController
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

@Configuration
@EnableMethodSecurity(prePostEnabled = true)
@EnableWebSecurity
open class SecurityConfig(
) {

    @Bean
    open fun securityFilterChain(
        http: HttpSecurity,
        jwtAuthFilter: JwtAuthenticationFilterController,
    ): SecurityFilterChain {
        http
            .csrf { it.disable() }
            .cors { }
            .authorizeHttpRequests { auth ->
                auth
                    // endpoint nào được permitAll thì không cần token
                    .requestMatchers("/graphql").permitAll()
                    .requestMatchers("/api/auth/**").permitAll()
                    .requestMatchers("/error", "/error/**").permitAll()
                    .anyRequest().authenticated()
            }
            //không lưu phiên đăng nhập trên server. Mỗi request đều phải kèm token để xác thực lại.
            .sessionManagement { session ->
                session.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
            }
            // nếu như không có dòng này thì khi có 1 rq gửi tới dù có token đúng thì vẫn lỗi 403
            // vì jwtAuthFilter đảm nhiệm việc đọc header, validate token, set authen vào SecurityContextHolder
            //SecurityContext là “chỗ chứa” thông tin bảo mật của request hiện tại — quan trọng nhất là đối tượng Authentication (ai đang gọi, đã đăng nhập chưa, có quyền gì).
            //SecurityContext trống nên
            .addFilterBefore(jwtAuthFilter, UsernamePasswordAuthenticationFilter::class.java)
        return http.build()
    }

    @Bean
    open fun corsConfigurationSource(): CorsConfigurationSource {
        val c = CorsConfiguration().apply {
            // ĐỔI thành origin FE của bạn. Ví dụ: vue ở 8081
            allowedOrigins = listOf("http://localhost:5173", "http://localhost:8081")
            // hoặc nếu đổi port liên tục: allowedOriginPatterns = listOf("http://localhost:*")
            allowedMethods = listOf("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
            allowedHeaders = listOf("Authorization", "Content-Type", "X-CSRF-TOKEN")
            exposedHeaders = listOf("Authorization", "Location")
            allowCredentials = true
            maxAge = 3600
        }
        return UrlBasedCorsConfigurationSource().apply {
            registerCorsConfiguration("/**", c)
        }
    }
}
